### ISSUE TITLE
[pointer_interceptor] Intercept "on top" of a view

### ISSUE THREAD

**Issue by ditman** [flutter/flutter#121272]
The current pointer interceptor always intercepts by assuming that its children is going to be a well behaved flutter element.

However, it's been reported that it's more practical in some cases to stack a pointer interceptor on *top* of an HTMLElementView to disable its interaction completely.

Can the pointer_interceptor be configured to be positioned "on top" of its child, rather than behind it?

**Comment by ditman** [comment]
The current version of the interceptor is just a [Stack](https://github.com/flutter/packages/blob/main/packages/pointer_interceptor/lib/src/web.dart#L84-L91).

It seems doable to intercept "on top" by rearranging the order of the Stack that the widget proposes?

**Comment by jacobsimionato** [comment]
Would this proposal also solve https://github.com/flutter/flutter/issues/74618 which is affecting a Google app in prod?

**Comment by ditman** [comment]
@jacobsimionato mmmaybe, this would effectively "disable" clicks in any platform view that is wrapped by the widget, I think.

**Comment by wanjm** [comment]
> Would this proposal solve the issue that when a PopupMenuItem is overlaid over a HTMLElementView, clicks on the menu also propagate to the HTML view underneath, as described in #74618 ? This issue is affecting a Google app in prod, and it would be great to have a fix.

@ditman  maybe I didn't catch the problem;
in my opinion when someone want to make the PopupMenuItem work, it just need wrap the PopupMenuItem with PointerInterceptor;  not matter  pointer_interceptor is on or behind it child;

but when pointer_interceptor is on its child, when using PointerInterceptor with webview, it can disable all the webview; so there is many button on top webview, and I don't need webview to be interactive, which is more convenient

**Comment by ditman** [comment]
> @ditman maybe I didn't catch the problem;

@wanjm I think you understood correctly:

* Blocking as it is now is intended for Flutter elements that are overlaid on TOP of a HtmlElementView (imagine: A Floating Action Button on top of a Google Maps view)
* Blocking as proposed in this issue (Intercepting "on top") is intended to disable mouse interactions in the HtmlElementView (imagine: A Flutter Modal Widget on top of a Google Maps View).

In the second case, it's easier to just "cover" the HtmlElementView with one widget, than trying to figure out where to put possile PointerInterceptors all around the app (behind buttons, text inputs, the modal itself, the ModalBarrier... etc)

**Comment by wanjm** [comment]
Blocking as proposed (on top) can also solve the float button case; in this case, we just wrap FAB with PointerInterceptor; and I try tested in https://github.com/flutter/flutter/issues/142427#issuecomment-1963197381, please refer to the code in the comments;

**Comment by ditman** [comment]
> so I think on top can solve all problems and easier to use;

I'll give this a shot, you may be right, and without expanding the API.

**Comment by wanjm** [comment]
> > so I think on top can solve all problems and easier to use;
> 
> Sorry @wangm, but intercepting "on top" by default breaks the standard behavior. You can verify by running [the example app](https://github.com/flutter/packages/tree/main/packages/pointer_interceptor/pointer_interceptor_web/example) of the package after doing the change to the package.
> 
> You'll see that the "Works as expected" button doesn't work as expected :)
> 
> (It was worth a try though!)


If child was put on top, then there is no need to wrap "Works As Expected" with PointerInterceptor any more;

**Comment by wanjm** [comment]
Yes, I tested the code with my change , the button "Works as expected" doesn't work any more;

I have checked that the Dom tree  Z-Index order is unchanged no matter where I put the intercepting on top or not;

I am curious how does the mouse event flows, I thought it  is first transferred  to Dom and finally to body (which transfer to flutter);

another thing I found that  through "Never calls onPressed" can't trigger onPress callback; but it reacts to hover, which means that it can receive mouse move event; how can it get it.

**Comment by wanjm** [comment]
@ditman 
is there detailed event path that tells how the event flows?

**Comment by ditman** [comment]
@wanjm mouse events are handled by the web engine at the "view" level, start here:

* https://github.com/flutter/engine/blob/main/lib/web_ui/lib/src/engine/pointer_binding.dart#L97

Most of the code relies in the DOM ordering of layers. We don't mess with the z-index. Take into account that all "canvas" layers where flutter paints are `pointer-events: none`, so the event needs to reach all the way "down" to the root element of the view to be seen by the web engine, and sent to the framework.

I don't think there's a comprehensive guide on how pointer events work in Flutter web.

**Comment by wanjm** [comment]
@ditman thank you very much. 
currently , I still unable to understand why there is the difference between the Intercept is on top or down as flutter is always the last one to get the event;
I will try to read the code;

**Comment by ditman** [comment]
@wanjm imagine you want to build this UI:

![121272_desired-ui](https://github.com/flutter/flutter/assets/1255594/5c752098-f8e6-4480-b2cd-240c87d6ed1b)

With the current capabilities of the pointer interceptor (and flutter), this is what you can do:

![121272_beneath-buttons](https://github.com/flutter/flutter/assets/1255594/dfdd6a38-57f6-44fc-a3b6-040e8c4be45f)

(As you can see, clicks anywhere other than the Yes/No buttons will be handled by background map, even when the modal is open)

When intercepting on top, you wrap the map instead, so you end up with this:

![121272_covering-map](https://github.com/flutter/flutter/assets/1255594/e7fa09f0-d2a6-4247-888e-ddeb94616857)

Which I think does what everybody wants. Of course this does **not** fix all the cases, but that particular one where the background is a platform view becomes more convenient.

**Comment by wanjm** [comment]
@ditman  thank you so much;

**Comment by wanjm** [comment]
![wecom-temp-178921-faec9316754dbd666f85d3da75694090](https://github.com/flutter/flutter/assets/3118573/e42f3ce4-0437-410f-b1f6-9bc2d03b9c0c)
currently, I want to know why build this way doesn't work, does the flutter check the z-index of interceptor;

**Comment by ditman** [comment]
> why build this way doesn't work

I am not sure, haven't had time to debug it!

There might be some clues here too:

* https://github.com/flutter/engine/blob/main/lib/web_ui/lib/src/engine/canvaskit/overlay_scene_optimizer.dart#L155-L159

Since the `pointer_interceptor` is an "invisible" platform_view.
